File: repeated_calls/orchestrator/agents/cause_agent.py
"""Pre-built Semantic Kernel agent for determining a product issue cause."""

from semantic_kernel import Kernel
from semantic_kernel.agents import ChatCompletionAgent
from semantic_kernel.connectors.ai import FunctionChoiceBehavior
from semantic_kernel.connectors.ai.open_ai import AzureChatPromptExecutionSettings
from semantic_kernel.functions import KernelArguments
from semantic_kernel.agents import AzureAIAgent,AzureAIAgentSettings, AzureAIAgentThread
from repeated_calls.orchestrator.entities.structured_output import CauseResult
from azure.identity.aio import DefaultAzureCredential
from azure.ai.agents.models import (
    ResponseFormatJsonSchema,
    ResponseFormatJsonSchemaType,
)
from semantic_kernel.connectors.mcp import MCPSsePlugin
from settings import McpApiSettings
from repeated_calls.orchestrator.plugins.mcp_plugins import McpApiKeyPlugin

async def get_agent_response(instructions: str, userprompt:str, thread_id:str) -> str:
    """Agent for determining if the call is repeated or not."""
    # Define temperature and which functions the agent can use
    ai_agent_settings = AzureAIAgentSettings()
    mcp_settings = McpApiSettings()
    async with (
        DefaultAzureCredential() as creds,
        AzureAIAgent.create_client(credential=creds, 
                                    conn_str=ai_agent_settings.endpoint, 
                                    deployment_name=ai_agent_settings.model_deployment_name) as client,
        MCPSsePlugin(
            name="CustomerDataPlugin",
            description="Customer domain data and product related data",
            url=mcp_settings.customer_url
        ) as customer_plugin,
        MCPSsePlugin(
            name="OperationsDataPlugin",
            description="Operations data",
            url=mcp_settings.operations_url
        ) as operations_plugin
    ):
        # Create agent definition
        agent_definition = await client.agents.create_agent(
            model=ai_agent_settings.model_deployment_name,
            name="CauseAgent",
            instructions=instructions,
            response_format=ResponseFormatJsonSchemaType(
                json_schema=ResponseFormatJsonSchema(
                    name="cause",
                    description="Extract cause information.",
                    schema=CauseResult.model_json_schema(),
                )
            )
        )

        # Create the AzureAI Agent
        agent = AzureAIAgent(
            client=client,
            definition=agent_definition,
            plugins=[McpApiKeyPlugin(), customer_plugin, operations_plugin]
        )

        thread = AzureAIAgentThread(client=client, thread_id=thread_id)
        response = await agent.get_response(
            messages=userprompt,
            thread=thread
        )

        await client.agents.delete_agent(agent.id)

    return response.content

File: dotnet-samples/Agents/AzureAI/CauseAgent.cs
using Azure.AI.Agents.Persistent;
using Azure.Identity;
using cw_repeated_calls_dotnet.Agents.SemanticKernel;
using cw_repeated_calls_dotnet.Entities.States;
using cw_repeated_calls_dotnet.Entities.StructuredOutput;
using cw_repeated_calls_dotnet.PromptEngineering;
using Microsoft.SemanticKernel;
using Microsoft.SemanticKernel.Agents;
using Microsoft.SemanticKernel.Agents.AzureAI;
using Microsoft.SemanticKernel.ChatCompletion;
using Microsoft.SemanticKernel.Connectors.OpenAI;
using Microsoft.SemanticKernel.Services;
using ModelContextProtocol.Protocol;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace cw_repeated_calls_dotnet.Agents.AzureAI
{
    public class CauseAgent : IAgent<CauseResult>
    {
        public async Task<CauseResult> GetAgentResponseAsync(Kernel kernel, RepeatedCallState state)
        {
            Kernel agentKernel = kernel.Clone();
            var plugins = agentKernel.Plugins;

            var prompts = new CausePrompts(state);

            var client = agentKernel.GetRequiredService<PersistentAgentsClient>();
            var chatCompletion = kernel.GetRequiredService<IChatCompletionService>();
            string model = chatCompletion.Attributes["DeploymentName"].ToString();

            var definition = await client.Administration.CreateAgentAsync(
               model: model,
               name: "CauseAgent",
               instructions: prompts.SystemPrompt,
               responseFormat: BinaryData.FromString(JsonSerializer.Serialize(new
               {
                   type = "json_schema",
                   json_schema = new
                   {
                       name = "cause",
                       description = "Extract cause information.",
                       schema = new
                       {
                           type = "object",
                           properties = new
                           {
                               productId = new { type = "string", description = "The ID of the product causing the issue" },
                               customerId = new { type = "integer", description = "The ID of the customer" },
                               analysis = new { type = "string", description = "Analysis of the problem" },
                               conclusion = new { type = "string", description = "Conclusion of the analysis" },
                               isOperationsCause = new { type = "boolean", description = "Whether the issue is caused by operations" }
                           },
                           required = new[] { "productId", "customerId", "analysis", "conclusion", "isOperationsCause" }
                       }
                   }
               }))
               );

            AzureAIAgent causeAgent = new(definition, client, plugins:agentKernel.Plugins)
            {
                
                Arguments = new KernelArguments(
                    new OpenAIPromptExecutionSettings()
                    {
                        Temperature = 0,
                        FunctionChoiceBehavior = FunctionChoiceBehavior.Auto(options: new() { RetainArgumentTypes = true }),
                    })
            };

            AzureAIAgentThread thread = new AzureAIAgentThread(client, state.ThreadId);
            ChatMessageContent message = new(AuthorRole.User, prompts.UserPrompt);

            CauseResult causeResult = null;

            try
            {
                ChatMessageContent response = await causeAgent.InvokeAsync(message, thread: thread).FirstAsync();
                causeResult = JsonSerializer.Deserialize<CauseResult>(response!.ToString());
                
            }
            finally
            {
                await client.Administration.DeleteAgentAsync(causeAgent.Id);
            }
            
            if (causeResult == null)
            {
                throw new InvalidOperationException("Failed to deserialize the response from the Cause Agent.");
            }
            return causeResult;
        }
    }
}
